sat/apiclient/pcs.py

```diff
@@ -0,0 +1,193 @@
+                    err_msg: the error message
+                    xname: the actual xname which failed
+        """
+        self.message = message
+        self.xname_errs = xname_errs if xname_errs is not None else []
+        self.xnames = [xname_err['xname'] for xname_err in self.xname_errs
+                       if 'xname' in xname_err]
+
+    def __str__(self):
+        """Convert to str."""
+        if not self.xname_errs:
+            return self.message
+        else:
+            # A mapping from a tuple of (err_code, err_msg) to a list of xnames
+            # with that combination of err_code and err_msg.
+            xnames_by_err = defaultdict(list)
+            for xname_err in self.xname_errs:
+                xnames_by_err[(xname_err.get('e'), xname_err.get('err_msg'))].append(xname_err.get('xname'))
+
+            xname_err_summary = '\n'.join([f'xname(s) ({", ".join(xnames)}) failed with '
+                                           f'e={err_info[0]} and err_msg="{err_info[1]}"'
+                                           for err_info, xnames in xnames_by_err.items()])
+
+            return f'{self.message}\n{xname_err_summary}'
+
+
+class PCSClient(APIGatewayClient):
+    """Client for the Power Control Service."""
+    base_resource_path = 'power-control/v1/'
+
+    def set_xnames_power_state(self, xnames, power_state, force=False, recursive=False, prereq=False):
+        """Set the power state of the given xnames.
+
+        Args:
+            xnames (list): the xnames (str) to perform the power operation
+                against.
+            power_state (str): the desired power state. Either "on" or "off".
+            force (bool): if True, disable checks and force the power operation.
+            recursive (bool): if True, power on component and its descendants.
+            prereq (bool): if True, power on component and its ancestors.
+
+        Returns:
+            None
+
+        Raises:
+            ValueError: if the given `power_state` is not one of 'on' or 'off'
+            PCSError: if the attempt to power on/off the given xnames with PCS
+                fails. This exception contains more specific information about
+                the failure, which will be included in its __str__.
+        """
+        allowed_states = {'on', 'off', 'soft-off', 'soft-restart', 'hard-restart', 'init', 'force-off'}
+        power_state = power_state.lower()
+        if power_state not in allowed_states:
+            allowed_states_str = ", ".join("\"" + state + "\"" for state in allowed_states)
+            raise ValueError(f'Invalid power state {power_state} given. Must be {allowed_states_str}')
+
+        if force:
+            if power_state in {'off', 'soft-off'}:
+                power_state = 'force-off'
+            elif power_state == 'soft-restart':
+                power_state = 'hard-restart'
+
+        target_xnames = set()
+        if recursive:
+            hsm_client = HSMClient(self.session)
```

Comment:
Yes, you are correct that `sat swap blade` should only be executed with the xname of a compute blade slot like `xXcCsS`. It looks like this `set_xnames_power_state` will be called with a list of HSM Node components in the slot and `recursive=True` in the case when the blade is an air-cooled (River) blade, and it will be called with the xname of the slot and `recursive=True` in the case when the blade is a liquid-cooled (Mountain) blade.

However, it would still be best to correctly implement the recursive behavior correctly for any type of component (at least for a Chassis xname, a ComputeModule/slot xname, a Node xname, and perhaps a RouterModule xname).

Reply:
Thanks for clarifying Ryan. Have added up accordingly the changes for it